### DISCUSSION TITLE
Using graph equality functions to test the Jlm/Mlir converter

### DISCUSSION THREAD

**Discussion by ProgrammingLouis** [phate/jlm#492] [Q&A] [answered]
The current Jlm/Mlir converter testing process is really tedious. It requires constructing a test graph in both Jlm and Mlir and then writing checks for each of the generated Jlm/Mlir representations.
So the idea is to have 2 functions that check for graph equality in Jlm rvsdg and in Mlir rvsdg.
Then, for each test, it will only be needed to write a function to construct the graph in Jlm and it's equivalent in Mlir, and run the equality function (either in Mlir for the backend or Jlm for the frontend) to pass the test.

The question is if such an equality check function already exists in jlm.
I'm also opened to suggestions.
@phate

**Comment by phate** [comment, marked as answer]
We used to have something for determining graph equivalence in [jive](https://github.com/phate/jive/blob/master/src/rvsdg/equivalence.cpp), but that was not ported over as part of the move to jlm. However, it should not be too difficult to write a method for the region class that checks for structural equivalence of two regions.

You say that you want to implement two methods, one on the jlm side and one on the mlir side. How about you only implement one method on the jlm side and then perform this in the unit tests:

1. You setup a JLM RVSDG in your unit test. Let's call it jlm-rvsdg-1.
2. You feed jlm-rvsdg-1 into the JLM -> MLIR conversion.
3. Then you feed the result of step 2 into the MLIR -> JLM conversion. Let's call the output jlm-rvsdg-2.
4. Then you check that jlm-rvsdg-1 and jlm-rvsdg-2 are structurally equivalent with the newly implemented method.

Would this not do the trick for you @ProgrammingLouis ?

**Reply by ProgrammingLouis** [reply]
@phate
Yes this would work to test that both backend and frontend parts of the convertor work but this does not indicate which of those caused an error if there is one.

**Reply by phate** [reply]
You can start with the equivalence method for jlm and see how you fare with it. If it does not fully work out, then you can always step it up and do the same on the MLIR side.
